src/kernels/kernelFinder.ts

```diff
@@ -0,0 +1,136 @@
+
+/**
+ * Generic class for finding kernels (both remote and local). Handles all of the caching of the results.
+ */
+@injectable()
+export class KernelFinder implements IKernelFinder {
+    private startTimeForFetching?: StopWatch;
+    private fetchingTelemetrySent = new Set<string>();
+    private _finders: IContributedKernelFinder[] = [];
+
+    constructor(
+        @inject(PreferredRemoteKernelIdProvider) private readonly preferredRemoteFinder: PreferredRemoteKernelIdProvider
+    ) {}
+
+    public registerKernelFinder(finder: IContributedKernelFinder) {
+        this._finders.push(finder);
+    }
+
+    @traceDecoratorVerbose('Rank Kernels', TraceOptions.BeforeCall | TraceOptions.Arguments)
+    @captureTelemetry(Telemetry.RankKernelsPerf)
+    public async rankKernels(
+        resource: Resource,
+        notebookMetadata?: nbformat.INotebookMetadata,
+        @logValue<PythonEnvironment>('uri') preferredInterpreter?: PythonEnvironment,
+        @ignoreLogging() cancelToken?: CancellationToken,
+        useCache?: 'useCache' | 'ignoreCache',
+        serverId?: string
+    ): Promise<KernelConnectionMetadata[] | undefined> {
+        try {
+            // Get list of all of the specs from the cache and without the cache (note, cached items will be validated before being returned)
+            let kernels = await this.listKernels(resource, cancelToken, useCache);
+            if (serverId) {
+                kernels = kernels.filter((kernel) => !isLocalConnection(kernel) && kernel.serverId === serverId);
+            }
+            const preferredRemoteKernelId =
+                resource &&
+                this.preferredRemoteFinder &&
+                this.preferredRemoteFinder.getPreferredRemoteKernelId(resource);
+
+            let rankedKernels = rankKernels(
+                kernels,
+                resource,
+                notebookMetadata,
+                preferredInterpreter,
+                preferredRemoteKernelId
+            );
+
+            return rankedKernels;
+        } catch (ex) {
+            traceError(`RankKernels crashed`, ex);
+            return undefined;
+        }
+    }
+
+    public async listKernels(
+        resource: Resource,
+        cancelToken: CancellationToken | undefined,
+        useCache: 'ignoreCache' | 'useCache' = 'ignoreCache'
+    ): Promise<KernelConnectionMetadata[]> {
+        this.startTimeForFetching = this.startTimeForFetching ?? new StopWatch();
+
+        const kernels: KernelConnectionMetadata[] = [];
+
+        const allKernels = await Promise.all(
+            this._finders.map((finder) => {
+                return finder
+                    .listContributedKernels(resource, cancelToken, useCache)
+                    .catch((ex) => {
+                        // Sometimes we can get errors from the socket level or jupyter, with the message 'Canceled', lets ignore those
+                        if (!isCancellationError(ex, true)) {
```

Comment:
This same catch chunk is also in the catch for listContributedKernels in both finders. Seems like it should only be here or in the finders, not both locations.